Unification of xref and anchor descriptions for proper synergy with the downstream documentation and providing a general content review

MichalMaler opened the pull request:
Closing https://issues.redhat.com/browse/QDOCS-175

The approach in which "naked" xrefs and anchors names/labels are automatically filled to xref/anchor rendered output, even thought this works competently fine in upstream docs, doesn't render properly for the single-sourced documentation (Panteon II limitation). 

Thus, I am proposing a change to always add desired names for xrefs and anchors.

This PR contains an addition to our contributor's guide that  this approach  needs to be applied when new docs are created.

This is another little step forward to a better synergy between Upstream and Downsream.

MichalMaler committed:
Adding the missing xref/anchors descriptions

Signed-off-by: Michal Maléř <mmaler@redhat.com>

Review

Signed-off-by: Michal Maléř <mmaler@redhat.com>

Review

Signed-off-by: Michal Maléř <mmaler@redhat.com>

Fixes2

Signed-off-by: Michal Maléř <mmaler@redhat.com>

Fixes3

Signed-off-by: Michal Maléř <mmaler@redhat.com>

Fixes4

Signed-off-by: Michal Maléř <mmaler@redhat.com>

Fixes5

Signed-off-by: Michal Maléř <mmaler@redhat.com>

MichalMaler commented:
@sberyozkin @gsmet Hello. I need to do a rebase and fix some conflicts (by applying my version), but problem was that the xref links I was tweaking were renamed. I will apply your changes back so that both, my tweaks, and your renaming are applied.

maxandersen commented:
Just to understand the implications here. The value inside [] is just the text used for the link?

This could potentially result in lots of duplicated names and if the title ever change having it be out of sync.

I'm wondering if not possible to merge in via an automated process? or possibly use adoc attributes in the label to avoid the repetition?

maxandersen commented:
so i just checked and could not find naked references.... this was already done repo wide in https://github.com/quarkusio/quarkus/pull/31548 ?

Not a fan this is required IMO but I guess we'll simply see how it goes.

Any ETA on when patreon system will get updated to not required this unnecessary duplication?

michelle-purcell commented on docs/src/main/asciidoc/doc-reference.adoc:
```diff
@@ -289,10 +288,10 @@ xref:<anchor-ID>[]
 --
 
-NOTE: The `[]` part of your `xref` anchors sets up a desired anchor label in cases where the default label does not fit the content.
+NOTE: The `[]` part of your `xref` anchors sets up a desired anchor label in cases where the default label does not fit the content. However, always specifying the desired anchor or `xref` name.
```
I agree with @maxandersen. It would be great to resolve this automatically because not all contributors read the guides and we have a few more issues like this in the PV tooling system.

However, if we don't find a smart way to resolve this, here's a minor tweak for clarity:

```suggestion
Important: To ensure that the inline anchor link renders correctly,  always populate the anchor label `[]` part of the`xref` with the correct text that fits the sentence.

MichalMaler replied:
Thank you for the suggestion. I am more for the original line because it is not an upstream issue they won't render elsewhere. That's why I wanted to say that they should specify them manually and check if the context matches its purpose and the displayed xref output fits the sentence, and it makes sense (looks good) overall. Or, ofc, keep it empty in case the heading that is automagically sourced there is great and rendered output of it fits the context and looks good in a sentence (and not causing an issue in other repo). :)

MichalMaler replied:
Something about these lines?

Important: The `[]` part of your `xref` anchors sets up a desired anchor name when the default name, the value of which `asciidoc` sources from chapter headings, file names, or anchors, does not fit in the current context. Always specify the intended anchor or `xref` name and verify if the displayed name fits the context of your content.

MichalMaler commented:
> Just to understand the implications here. The value inside [] is just the text used for the link?
> 
> This could potentially result in lots of duplicated names and if the title ever change having it be out of sync.
> 
> I'm wondering if not possible to merge in via an automated process? or possibly use adoc attributes in the label to avoid the repetition?

@maxandersen 
You are right, I was just manually adding what Upstream can fetch by itself, and yes, it is a great feature. In the future, when downstream will have better tooling, we can stop doing this. 

If a change of a heading appears, I (or the committer, ofc) can do a (non-capital sensitive) full-text search and replace the old (manually copy/pasted) xrefs labels with a value that mirrors the new heading, thus this is not a big issue, but again, you are right - it fixes something, but can produce an issue if committer just changes the heading without doing a check.

Right now, Upstream has other problems, such as **All In Caps Headings**, no-numbed lists for procedural steps, or content that was never reviewed and contains a waste of typos and other grammar/style errors. We are trying to fix that.

 https://github.com/quarkusio/quarkus/pull/31548  was my attempt to do the unification of used anchor macros and fix naked links that don't properly render in downstream. The problem here is, that we have 5 ways how to create an anchor ID, and we have 3 ways how to call it (Upstream has them all and their combinations), but only two combinations of those work in downstream. I am trying to fix that and make it united (sane results for downstream, but added functionality for downstream)
 
 What I did in both repositories:
 * anchors are most like done by this syntax : [[anchor]] 
 * then called by xref:anchor[Name of the anchor-most likely heading or table label]
 
 I also replaced <<>> way if doing anchors by xref macro.
 So no change for Upstream (accept added consistency and ton of fixes; minus that automatic xref update you mentioned), and added fixes and unification for downstream.
 
 So this is the follow-up of my first PR you mentioned.

I am for all tweaks or suggestions.

Pantheon II is pain, and it is unmaintained and abandoned. I ask for adding such tooling that would eliminate this or add the possibility of having G. Smet's multi-option snipped bars (Gradle, Maven, Quarkus CLI...), but it was denied.

The downstream repo will get a new tooling when applying the IBM tool-chain, and we already requested these features to be added there, but Stefan can answer more info about this.

MichalMaler commented:
To add to this, my PRs were touching 1/3 of all already created xrefs, thus if somebody will l change a heading, it would, in some cases, required a check on xrefs too, since most of them were already specified (manually).

michelle-purcell commented on docs/src/main/asciidoc/security-testing.adoc:
```diff
@@ -122,7 +122,7 @@ for example by setting `quarkus.http.auth.basic=true` or `%test.quarkus.http.aut
 
 You can also use Wiremock to mock the authorization OAuth2 and OIDC services:
-See xref:security-oauth2.adoc#integration-testing[OAuth2 Integration testing], xref:security-oidc-bearer-authentication-concept.adoc#integration-testing-wiremock[OpenID Connect Bearer Token Integration testing], xref:security-oidc-code-flow-authentication-concept.adoc#integration-testing-wiremock[OpenID Connect Authorization Code Flow Integration testing] and xref:security-jwt.adoc#integration-te ...
+See xref:security-oauth2.adoc#integration-testing[OAuth2 Integration testing], xref:security-oidc-bearer-authentication-concept.adoc#integration-testing-wiremock[OpenID Connect Bearer Token Integration testing], xref:security-oidc-code-flow-authentication-concept.adoc#integration-testing-wiremock[OpenID Connect Authorization Code Flow Integration testing] and xref:security-using-jwt-tutorial#inte ...
```
```suggestion
See xref:security-oauth2.adoc#integration-testing[OAuth2 Integration testing], xref:security-oidc-bearer-authentication-concept.adoc#integration-testing-wiremock[OpenID Connect Bearer Token Integration testing], xref:security-oidc-code-flow-authentication-concept.adoc#integration-testing-wiremock[OpenID Connect Authorization Code Flow Integration testing] and xref:security-using-jwt-tutorial.adoc#integration-testing-wiremock[SmallRye JWT Integration testing] for more details.
```

michelle-purcell commented on docs/src/main/asciidoc/security-testing.adoc:
```diff
@@ -89,7 +89,7 @@ This will run the test with an identity with the given username and roles. Note
 identity to be present.
 
-See xref:security-oidc-bearer-authentication-concept.adoc#integration-testing-security-annotation[OpenID Connect Bearer Token Integration testing], xref:security-oidc-code-flow-authentication-concept.adoc#integration-testing-security-annotation[OpenID Connect Authorization Code Flow Integration testing] and xref:security-jwt.adoc#integration-testing-security-annotation[SmallRye JWT Integration te ...
+See xref:security-oidc-bearer-authentication-concept.adoc#integration-testing-security-annotation[OpenID Connect Bearer Token Integration testing], xref:security-oidc-code-flow-authentication-concept.adoc#integration-testing-security-annotation[OpenID Connect Authorization Code Flow Integration testing] and xref:security-using-jwt-tutorial#integration-testing-security-annotation[SmallRye JWT Inte ...
```
```suggestion
See xref:security-oidc-bearer-authentication-concept.adoc#integration-testing-security-annotation[OpenID Connect Bearer Token Integration testing], xref:security-oidc-code-flow-authentication-concept.adoc#integration-testing-security-annotation[OpenID Connect Authorization Code Flow Integration testing] and xref:security-using-jwt-tutorial.adoc#integration-testing-security-annotation[SmallRye JWT Integration testing] for more details about testing the endpoint code which depends on the injected `JsonWebToken`.
```

sheilamjones commented on docs/src/main/asciidoc/qute-reference.adoc:
```diff
@@ -47,9 +50,9 @@ Qute.fmt("I am {#if ok}happy{#else}sad{/if}!", Map.of("ok", true)); <5>
 <4> Note that for a "text/html" template the special chars are replaced with html entities by default.
 <5> You can use any xref:basic-building-blocks[building block] in the template.
-In this case, the xref:if_section[] is used to render the appropriate part of the message based on the input data.
+In this case, the xref:if_section[If section] is used to render the appropriate part of the message based on the input data.
```
@MichalMaler: Just a question, should "section" be included as part of the xref name here? That is, is the name of the section "If" rather than "If section"?

MichalMaler replied:
Hello Sheila. Hope you are doing good.
The same file contains a section called "If Section". Adding "If section" description manually because thats the best way in my opinion how to call that chapter and its name, while taking in a consideration that headings in this doc don't follow the proper capitalization, and was just adding what appers there in docs when displayed on a portal. 
HOWEVER! :) Maybe a better context would be "if condition" instead of "if section", but I am not the author.

With this fix, I just apply what was displayed there anyway. And it is a good example of the fact that even anchors with [] empty brackets sometimes renders out of the context meanings, so even those should be checked from time to time. 
CC @maxandersen

gsmet reviewed (changes requested):
Not saying we won’t merge this in the end but I think we need to take a step back. 

Wouldn’t it be possible to post process our doc instead before inclusion in the other system rather than making our doc significantly less maintainable?

sberyozkin commented on docs/src/main/asciidoc/security-using-jwt-tutorial.adoc:
```diff
@@ -3,6 +3,7 @@ This guide is maintained in the main Quarkus repository
 and pull requests should be submitted there:
 https://github.com/quarkusio/quarkus/tree/main/docs/src/main/asciidoc
 ////
+[id="security-using-jwt-tutorial"]
```
@MichalMaler Can we deal with renaming this file as part of the future JWT doc refactoring ?

sberyozkin replied:
It just introduces a new variable, I'd rather us taking care of the whole smallrye-jwt doc refactoring in a dedicated step

MichalMaler replied:
Hello @sberyozkin Thank you for your feedback. 
This is not a renaming, I just add the ID that should be there since the file creation. Heading is the same. No work needs to be done, all works the same.  This file now just have an appropriate thing that should be there when a new file is created.

sberyozkin replied:
@MichalMaler Hi, I see the file name itself has changed, but it is not tutorial right now, it has everything in it, tutorial, reference. The file should remain as `security-jwt.adoc` for now IMHO

MichalMaler replied:
Hello @sberyozkin . I renamed it back, also provided little tweaks when I was in the file again. The rest will need to be done with complete re-factorization of the content. Is this OK in your books now?

inoxx03 commented:
@maxandersen I wish I had better answers for you, but...

> so i just checked and could not find naked references.... this was already done repo wide in #31548 ?

Yes, AFAIK @MichalMaler introduced the macro-formatted xrefs across all guide in his PR. This was necessary so that we can publish content form upstream on our downstream tooling. Not great, but this was by far the most practical solution

> Not a fan this is required IMO but I guess we'll simply see how it goes.

I hear you. Me neither, but without this links render wrong in PV2 at the moment.

> Any ETA on when patreon system will get updated to not required this duplication?

As to WHEN, I really don't know. I've been chasing new information about this for months, but beyond the odd "it might happen this year", I got nothing. And I'm not gonna pull estimates out of thin air.
However, Pantheon isn't just getting updated. It is going to be replaced by a completely different content publishing & management system. 
Myself, and a few other folks in CCS who are part of the advisory board for the tooling upgrade are pushing for standard AsciiDoctor support in the new system to get rid of exceptions like these. 
Non-standard markup support is annoying everyone, especially docs folks :)

sberyozkin commented on docs/src/main/asciidoc/security-authorize-web-endpoints-reference.adoc:
```diff
@@ -179,7 +179,7 @@ This cannot be combined with `deny-unannotated-endpoints`, as `deny` takes the e
 `quarkus.security.deny-unannotated-members=true|false`::
 - if set to true, the access will be denied to all CDI methods
-and Jakarta REST endpoints that do not have security annotations but are defined in classes that contain methods with
+and JAX-RS endpoints that do not have security annotations but are defined in classes that contain methods with
```
@MichalMaler I think it is fine, as `Jakarta` is an implementation detail, Guillaume, let me know please if it is of concern, I'm not sure we need to use `Jakarta` in such contexts

MichalMaler replied:
OK, will keep it as it is if @gsmet agrees with you on this. Dont want to interfere with anybody elses intention. But these changes should be hand over to docs team since it affects our work (in two repositories at the same time).

gsmet replied:
No it is not fine. I did a PR specifically to change that so I would appreciate if it wasn't lost.

MichalMaler replied:
It will be back as I said earlier (if this will be merged), dont worry @gsmet .
I just thought that we should know about these renaming changes, thats all.

sheilamjones commented on docs/src/main/asciidoc/security-authentication-mechanisms-concept.adoc:
```diff
@@ -27,7 +27,7 @@ The following table maps specific authentication requirements to a supported mec
 |Username and password |xref:security-basic-authentication-concept.adoc[Basic], xref:security-authentication-mechanisms-concept.adoc#form-auth[Form]
 
-|Bearer access token |xref:security-oidc-bearer-authentication-concept.adoc[OIDC Bearer], xref:security-jwt.adoc[JWT], xref:security-oauth2.adoc[OAuth2]
+|Bearer access token |xref:security-oidc-bearer-authentication-concept.adoc[OIDC Bearer], xref:security-using-jwt-tutorial[JWT], xref:security-oauth2.adoc[OAuth2]
```
Question @MichalMaler: I think the JWT file was one of the ones you mentioned earlier that was renamed. When I click on this link xref:security-using-jwt-tutorial[JWT], it doesn't open.

sheilamjones commented on docs/src/main/asciidoc/security-authentication-mechanisms-concept.adoc:
```diff
@@ -237,15 +240,15 @@ The `quarkus-smallrye-jwt` extension provides a MicroProfile JSON Web Token (JWT
 `quarkus-smallrye-jwt` also provides the JWT generation API, which you can use to easily create `signed`, `inner-signed`, and `encrypted` `JWT` tokens.
 
-For more information, see xref:security-jwt.adoc[Using SmallRye JWT role-based access control].
+For more information, see xref:security-using-jwt-tutorial[Using SmallRye JWT role-based access control].
```
Same as previous comment. This link doesn't work.

inoxx03 commented:
@maxandersen 

> Just to understand the implications here. The value inside [] is just the text used for the link?
 
Yes, the part in the parentheses (`[]`) at the end of the macro is the caption so just the text used for the link.

> This could potentially result in lots of duplicated names and if the title ever change having it be out of sync.
Yes, that is a risk


> I'm wondering if not possible to merge in via an automated process? or possibly use adoc attributes in the label to avoid the repetition?

Good Catch! Yes, I think we should be able to do that. we've do it on product docs before.
AsciiDoc attribute with the cation still has to be defined somewhere, though.

sheilamjones commented on docs/src/main/asciidoc/security-oidc-bearer-token-authentication-tutorial.adoc:
```diff
@@ -177,9 +177,15 @@ For more information, see xref:security-openid-connect-oidc-configuration-proper
+[NOTE]
+====
+Do not start the Keycloak server when you run the application in a dev mode - `Dev Services for Keycloak` will launch a container.
+See the <<keycloak-dev-mode, Running the Application in Dev mode>> section below for more information.
```
@MichalMaler: Think this link update was missed. Update to 
xref:keycloak-dev-model[Run the application in Dev mode ]

MichalMaler replied:
Just adding the context. This was not a scope of this PR, and, it works as I was sharing on slack and saying during our two recent calls:
![image](https://user-images.githubusercontent.com/48474054/224276901-39ada279-a99b-42f2-9dd2-691b9cf935f2.png)
Anchor was fix in my previous PR. Render properly in downstream too.

MichalMaler replied:
But I can add this for the sake of consistency :)

MichalMaler replied:
CC: @inoxx03 
Hello Stefan, this is the example of what is working on P2 (as our today's call discussion)

sheilamjones commented on docs/src/main/asciidoc/security-oauth2.adoc:
```diff
@@ -16,7 +16,7 @@ It can be used to implement an application authentication mechanism based on tok
 This extension provides a light-weight support for using the opaque Bearer Tokens and validating them by calling an introspection endpoint.
 
-If the OAuth2 Authentication server provides JWT Bearer Tokens then you should consider using either xref:security-oidc-bearer-authentication-concept.adoc[OIDC Bearer authentication] or xref:security-jwt.adoc[SmallRye JWT] extensions instead.
+If the OAuth2 Authentication server provides JWT Bearer Tokens then you should consider using either xref:security-oidc-bearer-authentication-concept.adoc[OIDC Bearer authentication] or xref:security-using-jwt-tutorial[SmallRye JWT] extensions instead.
```
Missing .adoc. Update to xref:security-using-jwt-tutorial.adoc[SmallRye JWT]

MichalMaler replied:
Good catch. Fixing the fix in all the docs now.

MichalMaler commented on docs/src/main/asciidoc/resteasy-reactive.adoc:
```diff
@@ -426,8 +426,8 @@ public class Endpoint {
-Any method parameter with no annotation will receive the method body.footnote:[Unless it is a
-xref:path-parameter[URI template parameter] or a xref:context-objects[context object].], after it has been mapped from
+Any method parameter with no annotation will receive the method body footnote:[Unless it is a
+xref:path-parameter[URI template parameter] or a xref:context-objects[context object]] after being mapped from
```
Hello @michelle-purcell This is not a typo, but a valid syntax for the footnote. 
footnote:**[**xref:ONE[]  or a xref:TWO[]**]**
This change would broke the footnote.
![image](https://user-images.githubusercontent.com/48474054/224294107-1cc5529a-8262-447d-bd6a-3e99b198b07d.png)

gsmet commented:
@ebullient I think we need you on this one. I really think we shouldn't make our work more difficult upstream because a tool is not up to date downstream. The `<<>>` -> `xref:` change was already borderline but this one will make things extremely difficult in the long term.
Especially since we have a lot of people writing AsciiDoc and I'm pretty sure we will miss some occurrences from time to time.

I have no idea what the workflow is but it doesn't sound that hard to have a script that postprocesses our doc to transform `<<>>` references to AsciiDoc that can be understood by the downstream tool (and at some point, if the limitations are removed, we could just drop the script).

That would be better for upstream and make things significantly more reliable for downstream.

ebullient commented:
> I think we need you on this one. I really think we shouldn't make our work more difficult upstream because a tool is not up to date downstream. 

It could be out of date, or it could just be a holdover from how docs are typically pulled apart and re-stitched together, such that <<>> references are not consistently resolvable. Nuance, I know, but I can see a few different reasons why the limitation would be there.

> The <<>> -> xref: change was already borderline but this one will make things extremely difficult in the long term. Especially since we have a lot of people writing AsciiDoc and I'm pretty sure we will miss some occurrences from time to time.

I do agree on this point... I usually use xref myself, but I don't put the exact title in the link target (usually), I use text that makes sense, and I guess I personally find xref easier to read (because I go back and forth to markdown). But that's definitely personal preference, and it won't be universal. <<>> is standard asciidoc syntax, so I expect we will be playing whack-a-mole here.

> I have no idea what the workflow is but it doesn't sound that hard to have a script that postprocesses our doc to transform <<>> references to AsciiDoc that can be understood by the downstream tool (and at some point, if the limitations are removed, we could just drop the script).

This was also mentioned by @maxandersen as an option, and @inoxx03 said doable, but maybe extra steps. If this is something we can help with, then I think it lightens the load for all concerned (in my opinion, anyway).  We should sit down with @inoxx03 and get a better idea of how/where/when such a script would run...

gsmet commented:
Converted to draft for now so that it doesn't get merged accidentally.

MichalMaler commented:
@gsmet If I re-apply your **JPA->Jakar** changes (all of them) and make those **[]**-xref parts empty again, will you be for merging?
If so, here is what we will get:
1) Current unification in anchors and xrefs for both repos, even though we will not insist on stopping of the <<>> format. As I found out, it is possible to use this style in downstream if we use **[[anchord]]** style for anchor creation (even thought it doesn't solve the problem of naked xrefs).
2) Lot of small grammar/style fixes + since I will do the work in those file again, I can provide a little deeper review, thus you will get even better upstream docs and I will have a better usage of my time.
3) First upstream/downstream victory

CC @inoxx03

MichalMaler commented:
@gsmet @ sberyozkin Reverted the changes and applied your suggestions. Also clean some of the [] xrefs already. If you will give me a green light, I will clean the rest of it, so that only grammar/style updates can go in.

sberyozkin reviewed (approved):
LGTM, thanks @MichalMaler, we can merge once the change request is lifted as well

gsmet commented:
Superseded by https://github.com/quarkusio/quarkus/pull/31981 where I incorporated the good changes of this PR.

Thanks!
